File: sleepy-test/java-junit/AsyncRequestQueueTests_AsyncRequestQueueWithBlockBehavior.java
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class AsyncRequestQueueTests {

    @Test
    public void asyncRequestQueueWithBlockBehavior() throws Exception {
        AsyncRequestQueue queue = new AsyncRequestQueue(10, AsyncTargetWrapperOverflowAction.BLOCK);
        CountDownLatch producerFinished = new CountDownLatch(1);
        int[] pushingEvent = {0};

        ExecutorService executor = Executors.newSingleThreadExecutor();
        executor.submit(() -> {
            try {
                for (int i = 0; i < 1000; ++i) {
                    AsyncLogEventInfo logEvent = LogEventInfo.createNullEvent().withContinuation(ex -> {});
                    logEvent.getLogEvent().setMessage("msg" + i);
                    pushingEvent[0] = i;
                    queue.enqueue(logEvent);
                }
            } finally {
                producerFinished.countDown();
            }
        });

        AsyncLogEventInfo[] logEventInfos;
        int total = 0;

        while (total < 500) {
            int left = 500 - total;
            logEventInfos = queue.dequeueBatch(left);
            int got = logEventInfos.length;
            assertTrue(got <= queue.getRequestLimit());
            total += got;
        }

        Thread.sleep(500);
        assertEquals(510, pushingEvent[0]);
        queue.dequeueBatch(1);
        total++;
        Thread.sleep(500);
        assertEquals(511, pushingEvent[0]);

        while (total < 1000) {
            int left = 1000 - total;
            logEventInfos = queue.dequeueBatch(left);
            int got = logEventInfos.length;
            assertTrue(got <= queue.getRequestLimit());
            total += got;
        }

        producerFinished.await();
        executor.shutdown();
    }
}

File: refactored/duplicate-assert/python-pytest/AsyncRequestQueueTests_AsyncRequestQueueWithBlockBehavior.py
import pytest
import time
from threading import Thread, Event

class TestAsyncRequestQueue:

    @pytest.mark.asyncio
    async def test_async_request_queue_with_block_behavior_1(self):
        queue = AsyncRequestQueue(10, AsyncTargetWrapperOverflowAction.Block)
        producer_finished = Event()
        pushing_event = [0]

        def producer_thread():
            for i in range(1000):
                log_event = LogEventInfo.create_null_event().with_continuation(lambda ex: None)
                log_event.get_log_event().set_message(f"msg{i}")
                pushing_event[0] = i
                queue.enqueue(log_event)
            producer_finished.set()

        producer = Thread(target=producer_thread)
        producer.start()

        total = 0
        while total < 500:
            left = 500 - total
            log_event_infos = queue.dequeue_batch(left)
            got = len(log_event_infos)
            assert got <= queue.get_request_limit()
            total += got

        time.sleep(0.5)
        assert pushing_event[0] == 510
        queue.dequeue_batch(1)
        total += 1
        time.sleep(0.5)
        assert pushing_event[0] == 511

        while total < 1000:
            left = 1000 - total
            log_event_infos = queue.dequeue_batch(left)
            got = len(log_event_infos)
            total += got

        producer_finished.wait()

    @pytest.mark.asyncio
    async def test_async_request_queue_with_block_behavior_2(self):
        queue = AsyncRequestQueue(10, AsyncTargetWrapperOverflowAction.Block)
        producer_finished = Event()
        pushing_event = [0]

        def producer_thread():
            for i in range(1000):
                log_event = LogEventInfo.create_null_event().with_continuation(lambda ex: None)
                log_event.get_log_event().set_message(f"msg{i}")
                pushing_event[0] = i
                queue.enqueue(log_event)
            producer_finished.set()

        producer = Thread(target=producer_thread)
        producer.start()

        total = 0
        while total < 500:
            left = 500 - total
            log_event_infos = queue.dequeue_batch(left)
            got = len(log_event_infos)
            total += got

        time.sleep(0.5)
        assert pushing_event[0] == 510
        queue.dequeue_batch(1)
        total += 1
        time.sleep(0.5)
        assert pushing_event[0] == 511

        while total < 1000:
            left = 1000 - total
            log_event_infos = queue.dequeue_batch(left)
            got = len(log_event_infos)
            assert got <= queue.get_request_limit()
            total += got

        producer_finished.wait()
